File: comp-math/NumericalIntegration/src/main/java/com/arslanka/numericalintegration/exceptions/ExceptionHandler.java
package com.arslanka.numericalintegration.exceptions;

import com.arslanka.numericalintegration.models.dtos.Response;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ControllerAdvice;

@ControllerAdvice
public class ExceptionHandler {

    @org.springframework.web.bind.annotation.ExceptionHandler(Exception.class)
    public ResponseEntity<Response> handleException(Exception exception) {
        Response response;
        if (exception instanceof BusinessException) {
            response = new Response(((BusinessException) exception).message());
        } else {
            response = new Response(HttpStatus.INTERNAL_SERVER_ERROR.getReasonPhrase());
        }
        return new ResponseEntity<>(response, HttpStatus.OK);
    }
}

File: web-basics/WebLab4/src/main/kotlin/com/example/arslanka/weblab4/exceptions/ExceptionHandler.kt
package com.example.arslanka.weblab4.exceptions

import com.example.arslanka.weblab4.uitls.Logging
import com.example.arslanka.weblab4.uitls.logger
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import javax.servlet.http.HttpServletRequest

@ControllerAdvice
class ExceptionHandler {

    companion object : Logging {
        val log = logger()
    }

    private val codeToStatus = mapOf( //TODO Custom exceptions with error codes and messages
        "1" to HttpStatus.BAD_REQUEST
    )

    @ExceptionHandler(value = [java.lang.Exception::class])
    fun onAnyException(ex: Exception, request: HttpServletRequest): ResponseEntity<io.tej.SwaggerCodgen.model.Error> {
        val error = io.tej.SwaggerCodgen.model.Error(code = "1", message = ex.message ?: "")
        val response =
            ResponseEntity(
                error,
                HttpStatus.UNAUTHORIZED
            )
        log.error(ex.toString()) { "${request.method} ${request.requestURI}, HTTP.${response.statusCode} " }
        return response
    }
}